builtins: Check max allowed points for st_generatepoints

mneverov opened the pull request:
builtins: Check max allowed points for st_generatepoints

Previously, the number of generated points was not checked that could lead to
OOM while creating too large slices. This patch adds a validation that checks
upper bound of the given parameter (number of points to return) as well as
number of the cells to generate.

Fixes: #59452

Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/59973)
<!-- Reviewable:end -->

otan commented on pkg/geo/geomfn/generate_points.go:
```diff
@@ -113,14 +120,22 @@ func generateRandomPointsFromPolygon(
 		sampleCellSize = bboxHeight / float64(sampleHeight)
 	}
+	n := sampleHeight * sampleWidth
+	if n > geo.MaxAllowedSplitPoints {
```
thanks for doing this!

i think this one is too restrictive. what we really care about is `iterations <= nPoints*10` below.
with this the area may be small but already out of range. i think we should remove this check and restore the check for `			"Requires too many iterations over grid",	`

otan replied:
I see we also shuffle `n` times -- i'd also restrict shuffling to `min(n, geo.MaxAllowedSplitPoints)`

otan replied:
oh i see, we'd have to shuffle a lot of points. i think we should still avoid this check though; the nPoints defence i think is good enough for now.

mneverov replied:
hi,
```
select st_asewkt(
    st_generatepoints(
        st_geomfromtext('POLYGON((0 0,0 100,0.00001 0.00000001,0.99999 0.00000001,1 100,1 0,0 0))'),
        500, 1996
    )
);
```
this command takes ~4 GB and at least a couple of minutes to complete even with the check (I aborted the command).
If you change 500 to 5000 the server crushes immediately, without any stacktrace output :(
With the check the error is:
`ERROR: st_generatepoints(): generating random points error: generated area is too large: 500014321, max 65336`.

I agree that the check is too restrictive. Maybe, I can change MaxAllowedSplitPoints check to some other sensible number?

otan replied:
I've seen cases in PostGIS returning max 100, so maybe 100 should be our max :)

sounds like a good idea. do you mind also adding this to the `info` section of the builtin?

mneverov replied:
Ok, will remove this check, change the validation in `GenerateRandomPoints`  (nPoints must be less than 100) and add this to the info section.
Will update the PR soon, thanks!

mneverov replied:
Actually, I found in [postgis](https://github.com/postgis/postgis/blob/08c2363c682ff4c602288adbb90e50a024c56849/liblwgeom/lwgeom_geos.c#L1796) the 100 check you mentioned.
Looks like postgis doesn't have MaxAllowedSplitPoints check. Does it make sense to remove the `MaxAllowedSplitPoints` check at all, and change [the validation](https://github.com/cockroachdb/cockroach/pull/59973/files#diff-ad843e286e45d0fa5184d8ad9b4b5d05e2a3ef32586d5ff44dc54e1f96bea4afR156) to:
```
for nPointsGenerated, iterations := 0, 0; nPointsGenerated < nPoints && iterations <= 100; iterations++ {
```
?

otan replied:
It would have to be before the points are generated /the shuffle but you can remove the check after that.

mneverov replied:
`It would have to be before the points are generated`
oh, yes, you're right.

It were too many suggestions from my side so now i'm confused about the next action :)
I still do this `Ok, will remove this check, change the validation in GenerateRandomPoints (nPoints must be less than 100) and add this to the info section.`, right?

otan replied:
Sorry yeah many ways to go about this. I think we can still get to a state where the npointsgenerated is huge, so if you want to do the iterations < 100 idea but before the shuffle I'd be ok. Maybe easier to just cap the number of points to generate to be 100, and if it is over, take a random sample of points without generating the whole array.

mneverov replied:
got it, thanks!
will update the PR

otan replied:
And also sorry for the back and forth. You've helped uncover some of the mystery and we appreciate that!

mneverov replied:
no problem!

Added a new constant `maxAllowedGridSize = 100 * geo.MaxAllowedSplitPoints` to limit the generated area size.
Why 100? Tbh chosen arbitrary based on the calculations: 
the max slice size will be 6,5M. The slice stores cells which are other slices, i.e. 24bytes header. So 6,5M * 24 = 156MB to instantiate the cells slice only. I think it is reasonable amount of memory  (although might be maybe increased twice?)

When I execute 
```
select st_asewkt(st_generatepoints(st_geomfromtext('POLYGON((0 0,0 100,0.00001 0.00000001,0.99999 0.00000001,1 100,1 0,0 0))'), 65, 1996));
```
the memory profiler shows 
![Screenshot from 2021-02-11 08-28-11](https://user-images.githubusercontent.com/1296281/107610950-20924e80-6c43-11eb-9568-e8fbec449657.png)

I mentioned the limit for the function argument in the function description, but not the 100 iterations limitation as postgis doesn't mentioned it either. Let me know if I need to add something like `The function may return less than requested number of points when calculation takes too long.`

mneverov committed:
builtins: Check max allowed points for st_generatepoints

Previously, the number of generated points was not checked that could lead to
OOM while creating too large slices. This patch adds a validation that checks
upper bound of the given parameter (number of points to return) as well as
number of the cells to generate.

Fixes: #59452

Release note: none

otan reviewed (approved):
thanks! for now I'm happy with it as is.

bors r+
